elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/AggregationDataStoreHelper.java

```diff
@@ -195,4 +216,94 @@ private void resolveMetricMap(Set<Metric> metrics) {
+                            String.format(
+                                    "Metric field %s must be aggregated before filtering in having clause.", field));
+                }
+            } else {
+                if (dimensionProjections.stream().noneMatch(dim -> dim.getName().equals(field))) {
+                    throw new InvalidOperationException(
+                            String.format(
+                                    "Dimension field %s must be grouped before filtering in having clause.", field));
+                }
+            }
+        } else if (havingClause instanceof AndFilterExpression) {
+            validateHavingClause(((AndFilterExpression) havingClause).getLeft());
+            validateHavingClause(((AndFilterExpression) havingClause).getRight());
+        } else if (havingClause instanceof OrFilterExpression) {
+            validateHavingClause(((OrFilterExpression) havingClause).getLeft());
+            validateHavingClause(((OrFilterExpression) havingClause).getRight());
+        } else if (havingClause instanceof NotFilterExpression) {
+            validateHavingClause(((NotFilterExpression) havingClause).getNegated());
+        }
+    }
+
+    private Set<String> getSortingFields(Sorting sorting) {
+        Set<Path.PathElement> sortingFields = new LinkedHashSet<>();
+        if (sorting != null) {
+            Map<Path, Sorting.SortOrder> sortClauses =
+                    sorting.getValidSortingRules(schema.getEntityClass(), dictionary);
+            sortClauses.entrySet().stream()
+                    .forEach((entry) -> {
+                        Path path = entry.getKey();
+                        //TODO add support for multiple level nested sorts
+                        if (path.getPathElements().size() > 2) {
+                            throw new UnsupportedOperationException(
+                                    "Currently sorting on double nested fields is not supported");
+                        }
+                        Path.PathElement currentElement = path.getPathElements().get(0);
+                        sortingFields.add(currentElement);
+            });
+        }
+        validateSorting(sortingFields);
+        return sortingFields.stream().map(sf -> sf.getFieldName()).collect(Collectors.toSet());
+    }
+
+    private void validateFieldStatus(Path.PathElement currentElement) {
+        String currentField = currentElement.getFieldName();
+        Class<?> currentClass = currentElement.getType();
+        if (schema.getMetrics().stream().anyMatch(m -> m.getName().contains(currentField))) {
+            if (metricMap.keySet().stream().noneMatch(m -> m.getName().equals(currentField))) {
+                // Metric not specified in query
+                throw new InvalidOperationException("Can't sort on metric that is not present in query");
+            }
+        }
+        if (dictionary.getIdFieldName(currentClass).equals(currentField)) {
```

Comment:
Also reject query for field name **id** which is a reserved keyword for all id fields.